cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll

```diff
@@ -0,0 +1,158 @@
+  exists(BinaryOperation oper, VariableAccess access |
+    (
+      oper.getOperator() = "<" or
+      oper.getOperator() = "<=" or
+      oper.getOperator() = ">" or
+      oper.getOperator() = ">="
+    ) and
+    oper.getLeftOperand() = access and
+    access.getTarget() = var and
+    // Comparing to 0 is not an upper bound check
+    not oper.getRightOperand().getValue() = "0"
+  )
+}
+
+private predicate instructionTaintStep(Instruction i1, Instruction i2) {
+  // Expressions computed from tainted data are also tainted
+  i2 = any(CallInstruction call |
+      isPureFunction(call.getStaticCallTarget().getName()) and
+      call.getAnArgument() = i1 and
+      forall(Instruction arg | arg = call.getAnArgument() | arg = i1 or predictableInstruction(arg)) and
+      // flow through `strlen` tends to cause dubious results, if the length is
+      // bounded.
+      not call.getStaticCallTarget().getName() = "strlen"
+    )
+  or
+  // Flow through pointer dereference
+  i2.(LoadInstruction).getSourceAddress() = i1
+  or
+  i2.(UnaryInstruction).getUnary() = i1
+  or
+  exists(BinaryInstruction bin |
+    bin = i2 and
+    predictableInstruction(i2.getAnOperand().getDef()) and
+    i1 = i2.getAnOperand().getDef()
+  )
+  // TODO: Check that we have flow from `a` to `a[i]`. It may work for constant
+  // `i` because there is flow through `predictable` `BinaryInstruction` and
+  // through `LoadInstruction`.
+  //
+  // TODO: Flow from argument to return of known functions: Port missing parts
+  // of `returnArgument` to the `interfaces.Taint` library.
+  //
+  // TODO: Flow from input argument to output argument of known functions: Port
+  // missing parts of `copyValueBetweenArguments` to the `interfaces.Taint`
+  // library and implement call side-effect nodes. This will help with the test
+  // for `ExecTainted.ql`. The test for `TaintedPath.ql` is more tricky because
+  // the output arg is a pointer addition expression.
+}
+
+predicate tainted(Expr source, Element tainted) {
+  exists(DefaultTaintTrackingCfg cfg, DataFlow::Node sink |
+    cfg.hasFlow(DataFlow::exprNode(source), sink)
+  |
+    sink.asExpr().getConversion*() = tainted
+    or
+    // For compatibility, send flow from arguments to parameters, even for
+    // functions with no body.
+    exists(FunctionCall call, int i |
+      sink.asExpr() = call.getArgument(i) and
+      tainted = resolveCall(call).getParameter(i)
+    )
+    or
+    // For compatibility, send flow into a `Variable` if there is flow to any
+    // Load or Store of that variable.
+    exists(CopyInstruction copy |
+      copy.getSourceValue() = sink.asInstruction() and
+      accessesVariable(copy, tainted) and
+      not hasUpperBoundsCheck(tainted)
+    )
+    or
+    // For compatibility, send flow into a `NotExpr` even if it's part of a
+    // short-circuiting condition and thus might get skipped.
+    tainted.(NotExpr).getOperand() = sink.asExpr()
```

Comment:
I'm confused about why this is necessary. Wouldn't this be a `LogicalNotInstruction` where `getUnaryOperand()` is either `sink` or a conversion of `sink`, and therefore be covered by `instructionTaintStep` and the default flow through `ConvertInstruction`? I don't think IR construction can omit the `LogicalNotInstruction`, and if it can, I'd consider that a bug

Reply:
I think it's deliberate that some `!`-expressions aren't translated directly. For example, consider the translation of the `!` in `if (!(e1 && e2))`? The control flow should be short-circuiting, so this just gets translated as if the `!` was pushed all the way in. See:

https://github.com/Semmle/ql/blob/b85823bec5d992b8db4e9b7f13a8d7f1db11677f/cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll#L167-L170

https://github.com/Semmle/ql/blob/b85823bec5d992b8db4e9b7f13a8d7f1db11677f/cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll#L245-L247

https://github.com/Semmle/ql/blob/b85823bec5d992b8db4e9b7f13a8d7f1db11677f/cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedCondition.qll#L32-L44

Reply:
I see. I assume that's to make analyzing the IR simpler, but I thought there was a guarantee that every evaluated `Expr` (i.e. ignoring arguments to `sizeof` etc.) is the `getAST` of some `Instruction`.